File: apps/ios/Graft/Models/TaskProgress.swift
import Foundation

enum TaskProgressStatus: String, Codable, Sendable {
    case pending, active, done
}

struct TaskProgressItem: Codable, Sendable, Equatable, Identifiable {
    let id: String
    let title: String
    let status: TaskProgressStatus
}

/// Unknown rich-event payloads must never prevent the surrounding event from decoding.
struct TimelineTaskData: Codable, Sendable, Equatable {
    struct Todo: Codable, Sendable, Equatable {
        enum Status: String, Codable, Sendable { case pending, in_progress, completed }
        let id: String
        let text: String
        let status: Status
    }

    let type: String
    var title: String? = nil
    var steps: [TaskProgressItem]? = nil
    var todos: [Todo]? = nil

    var items: [TaskProgressItem]? {
        let values: [TaskProgressItem]?
        switch type {
        case "plan": values = steps
        case "todo_update":
            values = todos?.map { todo in
                let status: TaskProgressStatus
                switch todo.status {
                case .pending: status = .pending
                case .in_progress: status = .active
                case .completed: status = .done
                }
                return TaskProgressItem(id: todo.id, title: todo.text, status: status)
            }
        default: values = nil
        }
        guard let values else { return nil }
        var ids: Set<String> = []
        let readable = values.prefix(100).filter {
            !$0.title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && ids.insert($0.id).inserted
        }
        return !values.isEmpty && readable.isEmpty ? nil : readable
    }
}

extension TimelineTaskData {
    init(from decoder: Decoder) throws {
        let container = try? decoder.container(keyedBy: CodingKeys.self)
        type = (try? container?.decode(String.self, forKey: .type)) ?? ""
        title = try? container?.decode(String.self, forKey: .title)
        steps = try? container?.decode([TaskProgressItem].self, forKey: .steps)
        todos = try? container?.decode([Todo].self, forKey: .todos)
    }
}

struct TaskProgress: Sendable, Equatable, Identifiable {
    let id: String
    let title: String
    let items: [TaskProgressItem]

    var completedCount: Int { items.filter { $0.status == .done }.count }
    var isComplete: Bool { items.allSatisfy { $0.status == .done } }
}

/// Task snapshots are independent of message arrival and streaming status.
/// Like desktop, unfinished work carries across turns until replaced or cleared.
struct TaskProgressState: Sendable, Equatable {
    private var latest: TaskProgress?
    private var current: TaskProgress?
    private var turnID: String?
    private var userIDs: Set<String> = []

    var visible: TaskProgress? {
        guard let list = current ?? latest, !list.items.isEmpty else { return nil }
        return current != nil || turnID == nil || !list.isComplete ? list : nil
    }

    mutating func beginTurn(id: String) {
        guard turnID != id else { return }
        turnID = id
        current = nil
    }

    mutating func fold(_ event: TimelineEvent) {
        if event.kind == "user.message" {
            if userIDs.insert(event.id).inserted { beginTurn(id: event.runId ?? event.id) }
            return
        }
        guard ["plan.update", "todo.update"].contains(event.kind),
              let items = event.data?.items else { return }
        let id = event.runId ?? turnID ?? event.id
        if turnID == nil { turnID = id }
        guard current == nil || id == turnID else { return }
        let title = event.data?.title?.trimmingCharacters(in: .whitespacesAndNewlines)
        let list = TaskProgress(id: id, title: (title?.isEmpty == false ? title : nil) ?? "Tasks", items: items)
        latest = list
        if id == turnID { current = list }
    }

    static func replay(_ events: [TimelineEvent]) -> Self {
        var state = Self()
        for event in events { state.fold(event) }
        return state
    }
}

File: apps/mobile-android/src/state/taskProgress.ts
import type { GraftTimelineEvent } from "@graft/mobile-contract";

export interface TaskProgressItem {
  readonly id: string;
  readonly title: string;
  readonly status: "pending" | "active" | "done";
}

export interface TaskProgress {
  readonly id: string;
  readonly title: string;
  readonly items: readonly TaskProgressItem[];
  readonly completedCount: number;
  readonly isComplete: boolean;
}

export function deriveTaskProgress(
  events: readonly GraftTimelineEvent[],
): TaskProgress | undefined {
  let turnId: string | undefined;
  let current: TaskProgress | undefined;
  let latest: TaskProgress | undefined;
  const users = new Set<string>();

  for (const event of events) {
    if (event.kind === "user.message") {
      if (!users.has(event.id)) {
        users.add(event.id);
        const nextTurn = event.runId ?? event.id;
        if (nextTurn !== turnId) {
          turnId = nextTurn;
          current = undefined;
        }
      }
      continue;
    }
    if (event.kind !== "plan.update" && event.kind !== "todo.update") continue;
    const data = event.data;
    let items: readonly TaskProgressItem[];
    if (data?.type === "plan") {
      items = data.steps;
    } else if (data?.type === "todo_update") {
      items = data.todos.map((todo) => ({
        id: todo.id,
        title: todo.text,
        status:
          todo.status === "completed"
            ? "done"
            : todo.status === "in_progress"
              ? "active"
              : "pending",
      }));
    } else {
      continue;
    }
    const ids = new Set<string>();
    const readable = items.filter((item) => {
      if (!item.title.trim() || ids.has(item.id)) return false;
      ids.add(item.id);
      return true;
    });
    if (items.length > 0 && readable.length === 0) continue;
    const id = event.runId ?? turnId ?? event.id;
    turnId ??= id;
    if (current && id !== turnId) continue;
    const completedCount = readable.filter((item) => item.status === "done").length;
    const list = {
      id,
      title: data.type === "plan" ? data.title?.trim() || "Tasks" : "Tasks",
      items: readable,
      completedCount,
      isComplete: completedCount === readable.length,
    };
    latest = list;
    if (id === turnId) current = list;
  }

  const selected = current ?? latest;
  if (!selected?.items.length) return undefined;
  return current || !turnId || !selected.isComplete ? selected : undefined;
}
